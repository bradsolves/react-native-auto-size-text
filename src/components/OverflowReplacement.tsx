import * as React from 'react';
import { NativeSyntheticEvent, Text, TextLayoutEventData } from 'react-native';

import { AutoSizeTextProps } from '../types';

const OverflowReplacement = (props: AutoSizeTextProps) => {
  const {
    fontSize,
    children,
    style,
    numberOfLines,
    overflowReplacement,
  } = props;
  const [currentText, setCurrentText] = React.useState<string>('');

  const handleResizing = (e: NativeSyntheticEvent<TextLayoutEventData>) => {
    const { lines } = e.nativeEvent;
    if (lines.length > (numberOfLines as number)) {
      setCurrentText(overflowReplacement as string);
      return;
    }

    setCurrentText(currentText);
  };

  return (
    <Text
      style={[
        style,
        {
          fontSize: fontSize,
        },
      ]}
      onTextLayout={handleResizing}
    >
      {currentText ? currentText : children}
    </Text>
  );
};

export default OverflowReplacement;
