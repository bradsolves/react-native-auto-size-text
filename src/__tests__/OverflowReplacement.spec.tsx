import { cleanup, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import OverflowReplacement from '../components/OverflowReplacement';

afterEach(cleanup);

describe('OverflowReplacement snapshot', () => {
  it('OverflowReplacement renders correctly', () => {
    const tree = renderer
      .create(
        <OverflowReplacement
          overflowReplacement='Overflowing'
          fontSize={14}
          numberOfLines={2}
        >
          Lorem impsum
        </OverflowReplacement>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('OverflowReplacement mode', () => {
  it('should pass the right props to Text component inside OverflowReplacement', async () => {
    const mockContent =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
    const mockNumberOfLines = 3;
    const mockFontSize = 24;
    const mockColor = 'red';

    const { getByTestId } = render(
      <OverflowReplacement
        style={{ color: mockColor }}
        fontSize={mockFontSize}
        numberOfLines={mockNumberOfLines}
        overflowReplacement={'foobar'}
      >
        {mockContent}
      </OverflowReplacement>
    );

    const textComponent = getByTestId('overflow-replacement');
    const content = textComponent.props.children;

    const numberOfLines = textComponent.props.numberOfLines;
    const adjustsFontSizeToFit = textComponent.props.adjustsFontSizeToFit;
    const fontSize = textComponent.props.style[1].fontSize;
    const color = textComponent.props.style[0].color;

    await waitFor(() => {
      expect(content).toBe(mockContent);
      expect(numberOfLines).toBeFalsy();
      expect(fontSize).toBe(mockFontSize);
      expect(color).toBe(mockColor);
      expect(adjustsFontSizeToFit).toBeFalsy();
    });
  });
});
