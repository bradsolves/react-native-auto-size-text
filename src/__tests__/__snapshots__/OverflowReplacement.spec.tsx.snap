// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowReplacement snapshot OverflowReplacement renders correctly 1`] = `
<Text
  onTextLayout={[Function]}
  style={
    Array [
      undefined,
      Object {
        "fontSize": 14,
      },
    ]
  }
  testID="overflow-replacement"
>
  Lorem impsum
</Text>
`;
